### PR TITLE
ROSbot 2R, XL webots_ros2 support

### PR DESCRIPTION
**Pull request by delihus**
**Description**
ROSbot 2R and ROSbot XL webots_ros2 support

**Related Topics**
https://github.com/cyberbotics/webots/pull/5973

**Affected Packages**
List of affected packages:
  - webots_ros2_husarion

### CODE REVIEW

**Comment by delihus** [comment]
I really don't want to copy the robots descriptions to the webots_ros2.
https://github.com/cyberbotics/webots_ros2/actions/runs/5220510624/jobs/9423668157?pr=770
I found an error with ROSbots.protos. I will try to fix this.
```bash
[robot_state_publisher-3] Error:   link 'solid' is not unique.
[robot_state_publisher-3]          at line 178 in ./urdf_parser/src/model.cpp
[robot_state_publisher-3] Failed to parse robot description using: urdf_xml_parser/URDFXMLParser
[robot_state_publisher-3] [WARN] [1686302208.945937669] [robot_state_publisher]: Unable to initialize urdf::model from robot description
```

**Comment by delihus** [comment]
> I really don't want to copy the robots descriptions to the webots_ros2. https://github.com/cyberbotics/webots_ros2/actions/runs/5220510624/jobs/9423668157?pr=770 I found an error with ROSbots.protos. I will try to fix this.
> 
> ```shell
> [robot_state_publisher-3] Error:   link 'solid' is not unique.
> [robot_state_publisher-3]          at line 178 in ./urdf_parser/src/model.cpp
> [robot_state_publisher-3] Failed to parse robot description using: urdf_xml_parser/URDFXMLParser
> [robot_state_publisher-3] [WARN] [1686302208.945937669] [robot_state_publisher]: Unable to initialize urdf::model from robot description
> ```

I found the problem. When I changed the name of `Solid` in `RpLidarA2.proto` in `Lidar.rotatingHead` to `rotating_head` I got the same error with another name:
```
[robot_state_publisher-3] Error:   link 'rotating_head' is not unique.
[robot_state_publisher-3]          at line 178 in ./urdf_parser/src/model.cpp
[robot_state_publisher-3] Failed to parse robot description using: urdf_xml_parser/URDFXMLParser
[robot_state_publisher-3] [WARN] [1686305352.530253249] [robot_state_publisher]: Unable to initialize urdf::model from robot description
```
When I delete whole `Lidar.rotatingHead` the error does not appear.  It looks like the `Lidar.rotatingHead` doubles the `Solid` tag.

**Comment by ygoumaz** [comment]
You are right, I think including the submodules is not useful and it would be cleaner to use the URDF exported by Webots from the PROTO. There is indeed a problem with the export of the `RPLidarA2` rotating head. I am checking the code of Webots to see if I can fix the bug before releasing the R2023b version. In the end, the cleanest solution would be to have a `webots_ros2_husarion` package with launch files and resources that you already added but without additional submodule components. In addition, a simple URDF file in the `resource` directory would allow to specify the ROS2 control joints and sensor topics (in the same way as https://github.com/cyberbotics/webots_ros2/blob/develop/webots_ros2_tiago/resource/tiago_webots.urdf).

**Comment by ygoumaz** [comment]
We are preparing the repository for the release of the next Webots version, so I rebased this PR to master, as we just merged develop into master (#779).

**Comment by ygoumaz** [comment]
The fix should be merged tomorrow, so you should be able to update the branch and remove the submodules.

**Comment by delihus** [comment]
As I see in https://github.com/cyberbotics/webots_ros2/actions/runs/5344139662/jobs/9688152381?pr=770 the `webots_ros2_husarion` is built but test cannot find the sources.

**Comment by ygoumaz** [comment]
What is the status of this PR? @delihus do you need any help on solving current test failures?

**Comment by delihus** [comment]
@ygoumaz 
1. I want to implement here your changes with proto links names replacement.
2. I cannot find out why the the action doesn't pass... 
https://github.com/cyberbotics/webots_ros2/actions/runs/5344139662/jobs/9688152381?pr=770

**Comment by ygoumaz** [comment]
1. is now ready in the sources as #789 was merged a few minutes ago.
2. I will try to help on this one.

**Comment by delihus** [comment]
I added remappings here [#d8e7e34](https://github.com/cyberbotics/webots_ros2/pull/770/commits/d8e7e34db9cbc81a2c90efc29299ff5333f5a95e) and it works great.

The last issue what I described you @ygoumaz in Discord. The converter from .proto to .urdf doesn't create collision from meshes. Only from boxes, cylinders.

Screenshot from 2023-07-05 10-18-53
But I think we can create the issue from that and pass this PR. What do you think @ygoumaz?

**Comment by delihus** [comment]
It looks like in the tests the `WebotsNode` doesn't remap links yet.
After: `launch_test src/webots_ros2/webots_ros2_tests/test/test_system_rosbot.py`
image
image

 I'm looking whats wrong...

**Comment by ygoumaz** [comment]
FYI you don't need double quotes in your components mapping file around keys with spaces. YAML file format allows to use spaces in key values.

Concerning the collision of meshes in RViz, I will investigate and agree that we should open another issue and address the problem in another PR!

**Comment by delihus** [comment]
So. From my side I added everything. Locally the tests with ROSbots passed.
ROSbot XL
image
ROSbot 2R
image
I don't know how to pass them on github.

**Comment by ygoumaz** [comment]
Nice! You should be close to having correct tests. I will check for you why they don't pass in the CI. I will have time for this tomorrow.

**Comment by ygoumaz** [comment]
The URLs in your world files are not in the correct format. They should target PROTOs on Github. This probably explains why the tests pass locally (installation from sources) but not in the CI (installation from binary release).

**Comment by delihus** [comment]
As now I see the test `testMovement` pass. The `testScan` does not work.  Maybe after my changes I described before?

File: webots_ros2_tests/test/test_system_rosbot_xl.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+            return False
+
+        self.wait_for_messages(self.__node, Odometry, '/odometry/filtered', condition=on_position_message_received)
+
+    def testScan(self):
+        def on_scan_message_received(message):
+            # There should be at least 1 range bigger than 0 and some = 0
+            number_of_inf = 0
+            number_of_non_zeroes = 0
+            for value in message.ranges:
```
[the remark this answers is not in the document]
Comment (delihus) [reply webots_ros2_tests/test/test_system_rosbot_xl.py:90]: What do you think @ygoumaz?

File: webots_ros2_tests/test/test_system_rosbot.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+        self.wait_for_messages(self.__node, Odometry, '/odometry/filtered', condition=on_position_message_received)
+
+    def testScan(self):
+        def on_scan_message_received(message):
+            # There should be at least 1 range bigger than 0 and some = 0
+            number_of_inf = 0
+            number_of_non_zeroes = 0
+            for value in message.ranges:
+                if value == float('inf'):
```
Comment (ygoumaz) [inline webots_ros2_tests/test/test_system_rosbot.py:90]: Concerning the Scan test, I believe your implementation is correct. The E-puck one, on the other side, isn't. The e-puck node is publishing on the `/scan` topic the values of DistanceSensors (and not a Lidar). The node should be updated to use `inf` as `OUT_OF_RANGE` value. I will open an issue.

**Review by ygoumaz — CHANGES_REQUESTED** [review]
I tested a bit more locally and still can't make the tests work on my side. I have given you some suggestions that fixed the tests on my machine. Let's try to apply them and see what comes out. There is probably some more review changes to do, but let's move on step-by-step. Making tests pass would already be a good start!

**Comment by ygoumaz** [comment]
> The last issue what I described you @ygoumaz in Discord. The converter from .proto to .urdf doesn't create collision from meshes. Only from boxes, cylinders.
> 
> Screenshot from 2023-07-05 10-18-53 But I think we can create the issue from that and pass this PR. What do you think @ygoumaz?

I have checked a little more the origin of this issue. In fact the robot model is constructed from the URDF generated by Webots and it is indeed not creating any visual if the bounding object is not a primitive shape. I don't think it would be too hard to implement the export of such Shapes. The ROSbot 2R is made only of primitives as BoundingObjects, so it is completely displayed in RViz. I will open an issue on the Webots repository about this enhancement.

File: webots_ros2_tests/test/test_system_rosbot.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            twist_message.linear.x = 0.5
+            twist_message.angular.z = 0.3
+            publisher.publish(twist_message)
+
+            # ROSbot should move in an arc to check the sensor fusion
+            if message.pose.pose.position.x > 0.5 and message.pose.pose.orientation.w < 0.9:
+                return True
+            return False
+
+        self.wait_for_messages(self.__node, Odometry, '/odometry/filtered', condition=on_position_message_received)
```
[the remark this answers is not in the document]
Comment (delihus) [reply webots_ros2_tests/test/test_system_rosbot.py:82]: I hope this is just a typo but the topic is the `/odometry/filtered` and you wrote `/odomtery/filtered`.
I can see the topic values:
```
ros2 topic echo /odometry/filtered --no-arr
```
```
header:
  stamp:
    sec: 129
    nanosec: 952000000
  frame_id: odom
child_frame_id: base_link
pose:
  pose:
    position:
      x: 1.0383662210156888
      y: 1.6213873866472988
      z: 0.0
    orientation:
      x: 0.0
      y: 0.0
      z: 0.8298143859614555
      w: 0.5580395011568737
  covariance: '<array type: double[36]>'
twist:
  twist:
    linear:
      x: -5.246486641376404e-11
      y: 0.0
      z: 0.0
    angular:
      x: 0.0
      y: 0.0
      z: 6.104560892043788e-09
  covariance: '<array type: double[36]>'
  ```
  
```
ros2 topic hz /odometry/filtered
```
gives:
```
average rate: 24.245
        min: 0.030s max: 0.067s std dev: 0.01411s window: 26
average rate: 24.083
        min: 0.030s max: 0.067s std dev: 0.01437s window: 50
average rate: 24.318
        min: 0.030s max: 0.068s std dev: 0.01443s window: 75
average rate: 24.437
        min: 0.030s max: 0.068s std dev: 0.01421s window: 100
```

File: webots_ros2_husarion/launch/rosbot_launch.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        ],
+        remappings=[
+            ('rosbot_base_controller/cmd_vel_unstamped', 'cmd_vel'),
+            ('odom', 'rosbot_base_controller/odom'),
+            ('rosbot/laser', '/scan'),
+            ('rosbot/rl_range', '/range/rl'),
+            ('rosbot/rr_range', '/range/rr'),
+            ('rosbot/fl_range', '/range/fl'),
+            ('rosbot/fr_range', '/range/fr')
+        ],
```
[the remark this answers is not in the document]
Comment (delihus) [reply webots_ros2_husarion/launch/rosbot_launch.py:85]: Sure. This remapping is not necessary. Probably I just forgot to remove this while testing.

FYI nav2 and slam_toolbox reads odometry from `/tf` not from topic and the frame is still `odom`.

File: webots_ros2_tests/test/test_system_rosbot.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (delihus) [reply webots_ros2_tests/test/test_system_rosbot.py:82]: Your all changes look like you skip the `robot_localization` which fuses the IMU and the encoders position and orientation. Maybe this is the dependency problem?

**Comment by delihus** [comment]
[260627f](https://github.com/cyberbotics/webots_ros2/pull/770/commits/260627f59c278f7e1234edc1db856d47faf607ac) this is a wrong lead because I see in the action logs that the `robot_localization` is installed.

**Comment by delihus** [comment]
All in all the `testScans` does not pass  [testScan (test_system_rosbot_xl.TestROSbotXL) ... FAIL](https://github.com/cyberbotics/webots_ros2/actions/runs/5476074626/jobs/9973086959#step:3:13669) unlike `testMovement`.

**Comment by delihus** [comment]
> All in all the `testScans` does not pass [testScan (test_system_rosbot_xl.TestROSbotXL) ... FAIL](https://github.com/cyberbotics/webots_ros2/actions/runs/5476074626/jobs/9973086959#step:3:13669) unlike `testMovement`.
It is kind of weird behavior. Not it is the other way round ( testMovement (test_system_rosbot_xl.TestROSbotXL) ... FAIL)[https://github.com/cyberbotics/webots_ros2/actions/runs/5516627742/jobs/10058205676?pr=770#step:3:15184].

**Comment by ygoumaz** [comment]
Maybe adding some debug logs to track the position of the robot could help understand the behavior of the robot and why the tests don't get the correct values.

**Comment by delihus** [comment]
When there  is only one test it pass as here [206679b](https://github.com/cyberbotics/webots_ros2/pull/770/commits/206679be985a8717ddfd7c750791c55165c36b0b).

**Comment by delihus** [comment]
The difference between local and remote build it that the implementation of the out of range in `LaserScan` locally is `inf` and remote is `nan`.

**Comment by ygoumaz** [comment]
So the `RPLidarA2` is giving `nan` values in output in the CI? Or is this a problem introduced by the Lidar plugin in the `webots_ros2_driver` package?

**Comment by delihus** [comment]
These are logs from github action and there are `inf` and `nans`. I think there should be no `nans`:
```python
print(message.ranges)
array('f', [nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, 0.217191681265831, 0.22657224535942078, 0.22516748309135437, 0.22383493185043335, 0.23317134380340576, 0.23192067444324493, 0.2412281483411789, 0.2505028247833252, 0.24936223030090332, 0.25863850116729736, 0.2575974464416504, 0.2668929398059845, 0.26595744490623474, 0.2752895951271057, 0.2846313416957855, 0.28385210037231445, 0.2932591438293457, 0.30269402265548706, 0.312163770198822, 0.32167527079582214, 0.3312353789806366, 0.34085118770599365, 0.34051448106765747, 0.3502700626850128, 0.37010273337364197, 0.3800116777420044, 0.3900119662284851, 0.4101138412952423, 0.42032408714294434, 0.4406657814979553, 0.4511265754699707, 0.47175952792167664, 0.49256524443626404, 0.5236295461654663, 0.5448492169380188, 0.576405942440033, 0.6082544326782227, 0.6404205560684204, 0.6831260323524475, 0.726268470287323, 0.7801486253738403, 0.8346157670021057, 0.9000619649887085, 0.9766687750816345, 1.0750747919082642, 1.3110227584838867, 1.4758588075637817, 1.695791482925415, 1.9825382232666016, 2.380342483520508, 2.967017889022827, nan, nan, inf, inf, inf, inf, inf, inf, inf, inf, inf, inf, inf, inf, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, 5.3751726150512695, 5.265248775482178, 5.180513381958008, 5.120338439941406, 5.073215961456299, 5.03884744644165, 5.0169572830200195, 4.9965739250183105, 4.988321304321289, 4.981387615203857, 4.975752830505371, 4.98188591003418, 5.02049446105957, 5.028804779052734, 5.038277626037598, 5.048910140991211, 5.0607008934021, 5.073650360107422, 5.087759971618652, 5.103033542633057, 5.1194748878479, 5.14720344543457, 5.16597843170166, 5.185946941375732, 5.207118034362793, 5.239541530609131, 5.263143539428711, 5.298004627227783, 5.3241047859191895, 5.361487865447998, 5.400166034698486, 5.430166721343994, 5.471518516540527, 5.514251708984375, 5.548382759094238, 5.593969345092773, 5.6410393714904785, 5.68963098526001, 5.739785194396973, 5.791545867919922, 5.8550705909729, 5.910205841064453, 5.967092990875244, 6.025783538818359, 6.096563339233398, 6.1590681076049805, 6.233858108520508, 6.300433158874512, 6.379517078399658, 6.460886478424072, 6.597066402435303, 6.683532238006592, 6.7725677490234375, 6.864272117614746, 6.9587483406066895, 6.915848731994629, 6.830679893493652, 6.6367292404174805, 6.573853492736816, 6.500601291656494, 6.439030170440674, 6.3893561363220215, 6.238186836242676, 6.1891398429870605, 6.152326583862305, 6.0229878425598145, 5.986962795257568, 5.868393421173096, 5.845071315765381, 5.822712421417236, 5.727569580078125, 5.718719959259033, 5.635613918304443, 5.56463098526001, 5.55773401260376, 5.486730575561523, 5.428438186645508, 5.437015056610107, 5.392726421356201, 5.348482131958008, nan, 5.262134552001953, 5.237918853759766, 5.20207405090332, 5.181944847106934, 5.137155532836914, 5.120948314666748, 5.106411457061768, 5.068017959594727, 5.056958198547363, 5.047393798828125, 5.014697551727295, 5.008263111114502, 4.979135990142822, 4.963733196258545, 4.961638450622559, 4.937450885772705, 4.9496169090271, 4.928389549255371, 4.920101165771484, 4.935692310333252, 4.918704032897949, 4.914277076721191, 4.900108814239502, 4.887343883514404, 4.919391632080078, 4.9090657234191895, 3.7243943214416504, 2.782052516937256, 2.217264413833618, 1.8412439823150635, 1.5803662538528442, 1.3723297119140625, 1.2165088653564453, 1.0921374559402466, 0.9888005256652832, 0.9062818884849548, 0.8343452215194702, 0.7728825807571411, 0.7218024730682373, 0.6810289621353149, 0.6404938101768494, 0.6001665592193604, 0.5700175166130066, 0.5400165915489197, 0.5201443433761597, 0.49037808179855347, 0.4707111716270447, 0.4511265754699707, 0.43160974979400635, 0.4221987724304199, 0.40279194712638855, 0.39350223541259766, 0.37415823340415955, 0.3649526834487915, 0.3557891845703125, 0.3466610014438629, 0.33756139874458313, 0.32848361134529114, 0.31942084431648254, 0.3103661835193634, 0.3013126850128174, 0.30269095301628113, 0.2831808924674988, 0.2740880846977234, 0.27556613087654114, 0.26647019386291504, 0.26805657148361206, 0.2589397430419922, 0.2606300711631775, 0.2514740824699402, 0.24225257337093353, 0.24404987692832947, 0.23475632071495056, 0.23664280772209167, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan])
```

Locally webots_ros2 echo:
```yaml
header:
  stamp:
    sec: 80
    nanosec: 896000000
  frame_id: laser
angle_min: 3.141592025756836
angle_max: -3.141592025756836
angle_increment: -0.015747329220175743
time_increment: 0.0
scan_time: 0.0
range_min: 0.20000000298023224
range_max: 12.0
ranges:
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- 0.2780417501926422
- 0.28191909193992615
- 0.2995957136154175
- 0.3178681433200836
- 0.3508087396621704
- 0.3788739740848541
- 0.4146113395690918
- 0.4493940770626068
- 0.5369895100593567
- 6.35524845123291
- 6.361996173858643
- 6.357226371765137
- 6.315924644470215
- 6.327488899230957
- 6.340324401855469
- 6.305247783660889
- 6.32111120223999
- 6.302095890045166
- 6.273114204406738
- 6.305171489715576
- 6.291457653045654
- 6.3257975578308105
- 6.303756237030029
- 6.30636739730835
- 6.344282627105713
- 6.338422775268555
- 6.3342061042785645
- 6.34261417388916
- 6.341522693634033
- 6.396297931671143
- 6.408761024475098
- 6.422635078430176
- 6.441125869750977
- 6.472313404083252
- 6.483658313751221
- 6.507784366607666
- 6.544974327087402
- 6.58435583114624
- 6.614729881286621
- 6.635887145996094
- 6.681978702545166
- 6.718989372253418
- 6.758354187011719
- 6.800154209136963
- 6.832573413848877
- 6.891412258148193
- 6.953218936920166
- 6.993536949157715
```

The real RPLidar:
```yaml
header:
  stamp:
    sec: 1689083986
    nanosec: 600664010
  frame_id: laser
angle_min: -3.1241390705108643
angle_max: 3.1415927410125732
angle_increment: 0.01745329238474369
time_increment: 0.00031720675178803504
scan_time: 0.11387722194194794
range_min: 0.15000000596046448
range_max: 12.0
ranges:
- .inf
- 0.4612500071525574
- .inf
- 0.45625001192092896
- 0.45524999499320984
- .inf
- 0.4582499861717224
- .inf
- .inf
- 0.4702500104904175
- .inf
- 0.46274998784065247
- .inf
- 0.4404999911785126
- .inf
- .inf
- 0.5214999914169312
- .inf
- 0.49125000834465027
- .inf
- 0.5087500214576721
- 0.30649998784065247
- .inf
- 0.28824999928474426
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- 0.7724999785423279
- .inf
- 0.7582499980926514
- .inf
- 0.8159999847412109
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- .inf
- 0.14875000715255737
```

File: .gitmodules [hunk 5]
```diff
@@ -1,3 +1,3 @@
 [submodule "webots_ros2_importer/webots_ros2_importer/urdf2webots"]
 	path = webots_ros2_importer/webots_ros2_importer/urdf2webots
-	url = https://github.com/cyberbotics/urdf2webots.git
+	url = https://github.com/cyberbotics/urdf2webots.git
```
Comment (ygoumaz) [inline .gitmodules:3]:
```suggestion
	url = https://github.com/cyberbotics/urdf2webots.git

```

File: webots_ros2_husarion/launch/rosbot_launch.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+from webots_ros2_driver.webots_controller import WebotsController
+from webots_ros2_driver.wait_for_controller_connection import WaitForControllerConnection
+
+
+def get_ros2_nodes(*args):
+    package_dir = get_package_share_directory('webots_ros2_husarion')
+    rosbot_webots_xacro_path = os.path.join(
+        package_dir, 'resource', 'rosbot_webots.urdf')
+    links_remappings_file_path = os.path.join(
+        package_dir, 'resource', 'rosbot_links_remappings.yaml')
```
Comment (ygoumaz) [inline webots_ros2_husarion/launch/rosbot_launch.py:38]:
```suggestion
    robot_description_path = os.path.join(package_dir, 'resource', 'rosbot_webots.urdf')
    links_remappings_file_path = os.path.join(package_dir, 'resource', 'rosbot_links_remappings.yaml')
```

File: webots_ros2_husarion/launch/rosbot_launch.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+
+    ekf_config = os.path.join(package_dir, 'resource', 'ekf.yaml')
+
+    ros2_control_params = os.path.join(
+        package_dir, 'resource', 'rosbot_controllers.yaml')
```
Comment (ygoumaz) [inline webots_ros2_husarion/launch/rosbot_launch.py:43]:
```suggestion
    ros2_control_params = os.path.join(package_dir, 'resource', 'rosbot_controllers.yaml')
```

File: webots_ros2_husarion/launch/rosbot_launch.py [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+    )
+    joint_state_broadcaster_spawner = Node(
+        package='controller_manager',
+        executable='spawner',
+        output='screen',
+        prefix=controller_manager_prefix,
+        arguments=['joint_state_broadcaster'] + controller_manager_timeout,
+    )
+    ros_control_spawners = [
+        diff_drive_controller_spawner, joint_state_broadcaster_spawner]
```
Comment (ygoumaz) [inline webots_ros2_husarion/launch/rosbot_launch.py:64]:
```suggestion
    ros_control_spawners = [diff_drive_controller_spawner, joint_state_broadcaster_spawner]
```

File: webots_ros2_husarion/launch/rosbot_xl_launch.py [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+from webots_ros2_driver.wait_for_controller_connection import WaitForControllerConnection
+
+
+def get_ros2_nodes(*args):
+    package_dir = get_package_share_directory('webots_ros2_husarion')
+    rosbot_xl_webots_xacro_path = os.path.join(package_dir, 'resource', 'rosbot_webots.urdf')
+    laser_filter_config = os.path.join(package_dir, 'resource', 'laser_filter.yaml')
+    ekf_config = os.path.join(package_dir, 'resource', 'ekf.yaml')
+    links_remappings_file_path = os.path.join(
+        package_dir, 'resource', 'rosbot_links_remappings.yaml')
```
Comment (ygoumaz) [inline webots_ros2_husarion/launch/rosbot_xl_launch.py:39]:
```suggestion
    links_remappings_file_path = os.path.join(package_dir, 'resource', 'rosbot_links_remappings.yaml')
```

File: webots_ros2_husarion/resource/ekf.yaml [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+                                      0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    1e-9, 0.0,     0.0,     0.0,     0.0,    0.0,    0.0,
+                                      0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    1e-9,  0.0,     0.0,     0.0,    0.0,    0.0,
+                                      0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,     1e-9,  0.0,     0.0,    0.0,    0.0,
+                                      0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,     0.0,     1e-9,  0.0,    0.0,    0.0,
+                                      0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,     0.0,     0.0,     1e-9, 0.0,    0.0,
+                                      0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,     0.0,     0.0,     0.0,    1e-9, 0.0,
+                                      0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,     0.0,     0.0,     0.0,    0.0,    1e-9]
```
Comment (ygoumaz) [inline webots_ros2_husarion/resource/ekf.yaml:84]:
```suggestion
                                      0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,    0.0,     0.0,     0.0,     0.0,    0.0,    1e-9]

```

File: webots_ros2_husarion/resource/rosbot_controllers.yaml [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+    angular:
+      z:
+        has_velocity_limits: true
+        has_acceleration_limits: true
+        has_jerk_limits: false
+        max_velocity: 3.0 # rad/s
+        # min_velocity -  When unspecified, -max_velocity is used
+        max_acceleration: 4.0 # rad/s^2
+        # min_acceleration - When unspecified, -max_acceleration is used.
+        max_jerk: 0.0 # rad/s^3
```
Comment (ygoumaz) [inline webots_ros2_husarion/resource/rosbot_controllers.yaml:47]:
```suggestion
        max_jerk: 0.0 # rad/s^3

```

File: webots_ros2_husarion/resource/rosbot_links_remappings.yaml [hunk 12]
```diff
@@ -0,0 +1,7 @@
+imu: imu_link
+cover: cover_link
+front left wheel: fl_wheel_link
+front right wheel: fr_wheel_link
+rear left wheel: rl_wheel_link
+rear right wheel: rr_wheel_link
+camera: camera_link
```
Comment (ygoumaz) [inline webots_ros2_husarion/resource/rosbot_links_remappings.yaml:7]:
```suggestion
camera: camera_link

```

File: webots_ros2_husarion/resource/rosbot_webots.urdf [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+            <state_interface name="position" />
+            <state_interface name="velocity" />
+        </joint>
+        <joint name="rr_wheel_joint">
+            <command_interface name="velocity" />
+            <state_interface name="position" />
+            <state_interface name="velocity" />
+        </joint>
+    </ros2_control>
+</robot>
```
Comment (ygoumaz) [inline webots_ros2_husarion/resource/rosbot_webots.urdf:42]:
```suggestion
</robot>

```

File: webots_ros2_husarion/resource/rosbot_xl_controllers.yaml [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+    angular:
+      z:
+        has_velocity_limits: true
+        has_acceleration_limits: true
+        has_jerk_limits: false
+        max_velocity: 3.0 # rad/s
+        # min_velocity -  When unspecified, -max_velocity is used
+        max_acceleration: 4.0 # rad/s^2
+        # min_acceleration - When unspecified, -max_acceleration is used.
+        max_jerk: 0.0 # rad/s^3
```
Comment (ygoumaz) [inline webots_ros2_husarion/resource/rosbot_xl_controllers.yaml:47]:
```suggestion
        max_jerk: 0.0 # rad/s^3

```

File: webots_ros2_husarion/launch/rosbot_launch.py [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+
+    rosbot_driver = WebotsController(
+        robot_name='rosbot',
+        parameters=[
+            {
+                'robot_description': rosbot_webots_xacro_path,
```
Comment (ygoumaz) [inline webots_ros2_husarion/launch/rosbot_launch.py:70]:
```suggestion
                'robot_description': robot_description_path,
```

File: webots_ros2_husarion/launch/rosbot_xl_launch.py [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+from launch_ros.actions import Node
+from ament_index_python.packages import get_package_share_directory
+from webots_ros2_driver.webots_launcher import WebotsLauncher
+from webots_ros2_driver.webots_controller import WebotsController
+from webots_ros2_driver.wait_for_controller_connection import WaitForControllerConnection
+
+
+def get_ros2_nodes(*args):
+    package_dir = get_package_share_directory('webots_ros2_husarion')
+    rosbot_xl_webots_xacro_path = os.path.join(package_dir, 'resource', 'rosbot_webots.urdf')
```
Comment (ygoumaz) [inline webots_ros2_husarion/launch/rosbot_xl_launch.py:35]:
```suggestion
    robot_description_path = os.path.join(package_dir, 'resource', 'rosbot_webots.urdf')
```

File: webots_ros2_husarion/launch/rosbot_xl_launch.py [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+        output='screen',
+        prefix=controller_manager_prefix,
+        arguments=['joint_state_broadcaster'] + controller_manager_timeout,
+    )
+    ros_control_spawners = [diff_drive_controller_spawner, joint_state_broadcaster_spawner]
+
+    rosbot_driver = WebotsController(
+        robot_name='rosbot_xl',
+        parameters=[
+            {'robot_description': rosbot_xl_webots_xacro_path,
```
Comment (ygoumaz) [inline webots_ros2_husarion/launch/rosbot_xl_launch.py:66]:
```suggestion
            {'robot_description': robot_description_path,
```

**Review by ygoumaz — CHANGES_REQUESTED** [review]
I just have a few minor suggestions on the code. Once https://github.com/cyberbotics/webots/pull/6324 is merged, it should be tested to confirm that Lidar messages are now correctly displayed.

File: webots_ros2_husarion/CHANGELOG.rst
Comment (ygoumaz) [inline webots_ros2_husarion/CHANGELOG.rst:1]: The changelog file should contain changes or at least mention the creation of the package (next version is 2023.1.2). The main changelog.rst file must be updated too.

**Review by ygoumaz — CHANGES_REQUESTED** [review]

File: webots_ros2_husarion/CHANGELOG.rst [hunk 18]
```diff
@@ -0,0 +1,7 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Changelog for package webots_ros2_husarion
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+2023.1.2 (2023-07-17)
```
Comment (ygoumaz) [inline webots_ros2_husarion/CHANGELOG.rst:5]:
```suggestion
2024.0.0 (2023-XX-XX)
```

**Review by ygoumaz — CHANGES_REQUESTED** [review]
Two last details:

- With the rebase, the version of the changelog becomes 2024.0.0
- The main changelog file of the package should have a new section 2024.0.0 and a new entry to explain the addition of the husarion sub-package: https://github.com/husarion/webots_ros2/blob/develop-husarion/webots_ros2/CHANGELOG.rst

**Pull request merged**